Browse links and "by regex" links in item display (Angular)

kshepherd opened the pull request:
## References
Related to [DSpace#8589](https://github.com/DSpace/DSpace/pull/8589)
Related to [REST Contract#205](https://github.com/DSpace/RestContract/pull/205)

## Description
This change allows the following configurable display of links in item page metdata values:
1. The configuration in `webui.browse.links.<n>` (dspace.cfg) now works as it did in previous versions, where a browse link is rendered for the metadata value in question if it is configured for a valid browse index. This works for both plain item page field components, and for metadata representations where there is virtual metadata instead of a link to a related entity. (related entity link will always take precedence)
3. The item page field components (and children) now allow a `[urlRegex]` declarative in the templates where a link will be rendered if and only if the regex matches the metadata text value, eg. `^https?://`, or something more restrictive

## To test conditional regex-based linking:

Try adding a section to a typed or untyped item template with a field like:
```
<ds-generic-item-page-field [item]="object"
      [fields]="['dc.subject']"
      [separator]="', '"
      [label]="'item.page.subject'">
      [urlRegex]="'^https?://www.something.com'"
    </ds-generic-item-page-field>
```

And try it out with different subject values, some of which match the regular expression.

## To test browse links:

Set up browse link config for fields, such as:
```
webui.browse.link.1 = author:dc.contributor.*
webui.browse.link.2 = subject:dc.subject.*
```

And try out items with values in these fields. With the author field, note that if it is not a "authorOf..." relation, the browse author link should be displayed. If it *is* a relation to a person entity, it should instead be linked to that entity.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

kshepherd commented:
Note for reviewers: I rebased this today (Jan 16 2023) after not touching since Nov 2022 - there were some pretty big changes to the shared module in that time but I believe I resolved the conflicts properly, and my testing is still successful.

tdonohue commented:
@kshepherd : Could you rebase this PR when you get a chance?  It has merge conflicts with latest `main`

kshepherd commented:
@tdonohue I have refactored this as requested. The /browses/ endpoint, controller and data services are now used. I also simplified some of the parent component level handling.

I am expecting that this PR will be constantly getting out of date with main as PRs are merged for 7.5. Is it OK if we continue to review, even if a rebase is needed? I don't expect any complicated conflicts. The only ones so far were rated to the new routeservice

tdonohue reviewed (commented):
@kshepherd : Reviewed & tested this today. As noted in the backend PR, I found everything to be working.  

However, there's some unused code in this PR (endpoints you implemented which don't use) which can be removed.  You also need to remove all hardcoded `<a href=` attributes, as you need to use `<a [routerLink]=` in Angular.  RouterLink ensures Angular can adapt the URLs if you deploy on a subpath... and it also ensures you aren't re-triggering SSR (If you look closely, right now your hardcoded "href" links all hit SSR instead of staying with CSR).

But, honestly, I think the cleanup here is relatively minor.  Overall this works and looks good & I think this should make it into 7.5 if you can get the updates made in the next week.  See comments inline below.

artlowel commented on src/app/item-page/simple/field-components/specific-field/generic/generic-item-page-field.component.ts:
```diff
@@ -40,5 +40,10 @@ export class GenericItemPageFieldComponent extends ItemPageFieldComponent {
+  /**
+   * Whether any valid HTTP(S) URL should be rendered as a link
+   */
+  @Input() urlRegex?;
```
Looks like you forgot the type here

kshepherd replied:
Thanks, I've resolved this.

artlowel reviewed (commented):
Thanks @kshepherd!

It works and the code is nearly there. Once @tdonohue's comments are addressed I'm a +1 for this PR

artlowel commented on src/app/core/browse/browse-link-data.service.ts:
```diff
@@ -0,0 +1,81 @@
+ * they are configured as browse links (webui.browse.link.<n>)
+ */
+@Injectable()
+export class BrowseLinkDataService extends IdentifiableDataService<BrowseDefinition> {
```
Hmm, looks like a review comment I made for the first version of this PR (but never submitted) was saved. This is no longer a problem since it was already addressed

kshepherd commented on src/app/shared/object-list/metadata-representation-list-element/plain-text/plain-text-metadata-list-element.component.html:
```diff
@@ -1,3 +1,16 @@
+  <span *ngIf="(metadataRepresentation.representationType=='authority_controlled')" class="dont-break-out">{{metadataRepresentation.getValue()}}</span>
+  <a *ngIf="(metadataRepresentation.representationType=='browse_link')"
+     class="dont-break-out ds-browse-link"
+     href="/browse/{{metadataRepresentation.browseDefinition.id}}?{{metadataRepresentation.browseDefinition.metadataBrowse?'value':'startsWith'}}={{metadataRepresentation.getValue()}}&bbm.page=1">
```
Thanks, this should be sorted.

kshepherd commented on src/app/shared/object-list/metadata-representation-list-element/plain-text/plain-text-metadata-list-element.component.html:
```diff
@@ -1,3 +1,16 @@
+    {{metadataRepresentation.getValue()}}
+  </span>
+  <a *ngIf="(metadataRepresentation.representationType=='plain_text') && isLink()" class="dont-break-out"
+  target="_blank" href="{{metadataRepresentation.getValue()}}">
```
I've used [href] here (note, not href but at least binding it properly) as this link could be external. From what I've seen, I think this is the intended usage?

kshepherd commented on src/app/shared/object-list/metadata-representation-list-element/metadata-representation-list-element.component.ts:
```diff
@@ -13,4 +13,11 @@ export class MetadataRepresentationListElementComponent {
    */
   metadataRepresentation: MetadataRepresentation;
+
+  isLink(): boolean {
```
Done

tdonohue replied:
Maybe I'm missing it, but I'm not seeing the new specs for this regex?

kshepherd commented on src/app/shared/object-list/metadata-representation-list-element/browse-link/browse-link-metadata-list-element.component.html:
```diff
@@ -0,0 +1,8 @@
+<div>
+  <a *ngIf="(metadataRepresentation.representationType=='browse_link')"
+     target="_blank" class="dont-break-out"
+     href="/browse/{{metadataRepresentation.browseDefinition.id}}?{{metadataRepresentation.browseDefinition.metadataBrowse?'value':'startsWith'}}={{metadataRepresentation.getValue()}}">
```
Thanks, this should be sorted.

kshepherd commented on src/app/item-page/field-components/metadata-values/metadata-values.component.ts:
```diff
@@ -40,12 +42,38 @@ export class MetadataValuesComponent implements OnChanges {
+   * Does this metadata value have a valid URL that should be rendered as a link?
+   * @param value
+   */
+  hasLink(value): boolean {
```
Done

kshepherd commented on src/app/core/browse/browse-definition-data.service.ts:
```diff
@@ -52,5 +68,109 @@ export class BrowseDefinitionDataService extends IdentifiableDataService<BrowseD
+   * @return {Observable<RemoteData<PaginatedList<T>>}
+   *    Return an observable that emits response from the server
+   */
+  public searchBy(searchMethod: string, options?: FindListOptions, useCachedVersionIfAvailable?: boolean, reRequestOnStale?: boolean, ...linksToFollow: FollowLinkConfig<BrowseDefinition>[]): Observable<RemoteData<PaginatedList<BrowseDefinition>>> {
```
Done

kshepherd commented on src/app/core/browse/browse-definition-data.service.ts:
```diff
@@ -21,8 +23,22 @@ import { dataService } from '../data/base/data-service.decorator';
   private findAllData: FindAllDataImpl<BrowseDefinition>;
+  private searchData: SearchDataImpl<BrowseDefinition>;
+
+  public static toSearchKeyArray(metadataKey: string): string[] {
```
I actually was able to remove this method

kshepherd commented on src/app/core/browse/browse-definition-data.service.ts:
```diff
@@ -52,5 +68,109 @@ export class BrowseDefinitionDataService extends IdentifiableDataService<BrowseD
+   * @param metadatumKey
+   * @param linkPath
+   */
+  findByFields(
```
Done

kshepherd commented on src/app/core/browse/browse-definition-data.service.ts:
```diff
@@ -52,5 +68,109 @@ export class BrowseDefinitionDataService extends IdentifiableDataService<BrowseD
+    );
+  }
+
+  findAllLinked(
```
Done

kshepherd commented on src/app/core/browse/browse-definition-data.service.ts:
```diff
@@ -52,5 +68,109 @@ export class BrowseDefinitionDataService extends IdentifiableDataService<BrowseD
+    return this.searchData.getSearchByHref(searchMethod, options, ...linksToFollow);
+  }
+
+  findByField(
```
Done.

kshepherd commented:
@tdonohue @artlowel I believe I have addressed all this feedback, thanks. Please let me know if there's anything further missing.

artlowel reviewed (approved):
Thanks @kshepherd!

tdonohue commented on src/app/item-page/field-components/metadata-values/metadata-values.component.html:
```diff
@@ -1,16 +1,37 @@
+<!-- Render value as a link to browse index -->
+<ng-template #browselink let-value="value">
+  <a class="dont-break-out preserve-line-breaks ds-browse-link"
+     href="/browse/{{browseDefinition.id}}?{{(browseDefinition.metadataBrowse?'value':'startsWith')}}={{value}}&bbm.page=1">
```
@kshepherd : I still think this should be a `routerLink` with `queryParams`.  Did that not work, or maybe this was just forgotten?

It appears you fixed this in `browse-link-metadata-list-element.component.html`, but not here.  I'm not sure exactly where this code is triggered though (it doesn't seem to be what appears on an Item page), but I'd expect this probably won't work if you deploy the UI on a subpath (e.g. http://localhost:4000/dspace/)

tdonohue commented on src/app/core/browse/browse.service.ts:
```diff
@@ -251,4 +251,37 @@ export class BrowseService {
+   * @param metadatumKey
+   * @param linkPath
+   */
+  getBrowseDefinitionFor(metadataKeys: string[]): Observable<BrowseDefinition> {
```
Also, I think we're missing specs here & updated TypeDocs (there's no "linkPath" param here).   While this isn't a complex method, I noticed that the `throw new Error` is commented out below...so we may want to write specs which show the expected before (should it throw an error? Or is it expected just to return undefined? (It looks like the latter, but the commented out code should then be removed)

tdonohue reviewed (changes requested):
@kshepherd : Overall, this works. But, there's a few prior inline comments that were overlooked.  My main concern here is the `routerLink` comment, as I think there's a bug in the current code (though I cannot figure out how to trigger that code to prove it).  Beyond that, there's just some missing specs that I had previously asked for... if it's too hard to create those specs, I can live without it. But ideally, we have specs for new methods.

kshepherd commented:
@tdonohue I've fixed up the link construction and some related tests, including isLink. Thanks for the review

tdonohue reviewed (approved):
👍 Thanks @kshepherd !  This now looks great to me. Verified it works for authors "out-of-the-box".  Also tested adding a browselink for Subject.   I appreciate your hard work in getting this ready to go.

tdonohue commented:
Merging as this is at +2
